src/main/java/org/spongepowered/api/schematic/Schematic.java

```diff
@@ -0,0 +1,210 @@
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.schematic;
+
+import com.flowpowered.math.vector.Vector3i;
+import com.google.common.collect.BiMap;
+import org.spongepowered.api.block.BlockSnapshot;
+import org.spongepowered.api.block.BlockState;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.api.data.DataSerializable;
+import org.spongepowered.api.data.persistence.DataBuilder;
+import org.spongepowered.api.data.persistence.InvalidDataException;
+import org.spongepowered.api.world.World;
+import org.spongepowered.api.world.extent.Extent;
+
+import java.util.Optional;
+
+/**
+ * Unless otherwise noted, "the specification" refers to the latest version of the Schematic Specification {@url https://github
+ * .com/SpongePowered/Schematic-Specification}.
+ */
+public interface Schematic extends DataSerializable {
+
+    /**
+     * Gets the attached {@link DataContainer}.
+     *
+     * <p>Correlates with the optional `Metadata` field of the specification.</p>
+     *
+     * @return The attached {@link DataContainer} or {@link Optional#empty()} if none
+     */
+    Optional<DataContainer> getMetadata();
+
+    /**
+     * Gets the width of the Schematic.
+     *
+     * <p>Correlates with the `Width` field of the specification.</p>
+     *
+     * @return The width of the Schematic.
+     */
+    int getWidth();
+
+    /**
+     * Gets the height of the Schematic.
+     *
+     * <p>Correlates with the `Height` field of the specification.</p>
+     *
+     * @return The height of the Schematic.
+     */
+    int getHeight();
+
+    /**
+     * Gets the length of the Schematic.
+     *
+     * <p>Correlates with the `Length` field of the specification.</p>
+     *
+     * @return The length of the Schematic.
+     */
+    int getLength();
+
+    /**
+     * Gets the origin of the Schematic (relative to the minimum point) or the minimum point if none.
+     *
+     * <p>Correlates with the `OffsetX`, `OffsetY`, and `OffsetZ` fields of the specification.</p>
+     *
+     * @return The origin of the Schematic
+     */
+    Vector3i getOrigin();
+
+    /**
+     * Gets the largest key in the palette.
+     *
+     * @return The largest key
+     */
+    int getPaletteMax();
+
+    /**
+     * Gets the palette of the Schematic.
+     *
+     * <p>Correlates backwards to the `Palette` field of the specification.</p>
+     *
+     * @return The palette
+     */
+    BiMap<Integer, BlockState> getPalette();
+
+    /**
+     * Gets the {@link BlockState} at the relative location, or {@link Optional#empty()} if none.
+     *
+     * @param extent The extent
+     * @param location The location (relative to the origin of the schematic) of the BlockSnapshot desired
+     * @return The BlockState found
+     */
+    Optional<BlockState> getBlockAt(Extent extent, Vector3i location);
+
+    /**
+     * Gets all {@link BlockState}s in the Schematic.
+     *
+     * @return All BlockStates
+     */
+    BlockState[] getBlocks();
```

Comment:
How about using an immutable list?
> On Apr 4, 2016, at 2:10 PM, Nick Condé <notifications@github.com> wrote:
> 
> In src/main/java/org/spongepowered/api/schematic/Schematic.java <https://github.com/SpongePowered/SpongeAPI/pull/1172#discussion_r58430610>:
> 
> > +
> > +    /**
> > +     * Gets the {@link BlockState} at the relative location, or {@link Optional#empty()} if none.
> > +     *
> > +     * @param extent The extent
> > +     * @param location The location (relative to the origin of the schematic) of the BlockSnapshot desired
> > +     * @return The BlockState found
> > +     */
> > +    Optional<BlockState> getBlockAt(Extent extent, Vector3i location);
> > +
> > +    /**
> > +     * Gets all {@link BlockState}s in the Schematic.
> > +     *
> > +     * @return All BlockStates
> > +     */
> > +    BlockState[] getBlocks();
> Well either the schematic is immutable in which case exposing arrays breaks that and allows you to mutate it willy nilly or its mutable in which case your enforcing the contract on all implementations that it is stored in memory as an array of blockstates (which in all honestly it probably wouldn't be as its cheaper to just keep it encoded to a byte/short/int array) better to just have getters and setters.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub <https://github.com/SpongePowered/SpongeAPI/pull/1172/files/3e69789b55ffb54141cfed6d82c3b58d07c707f5#r58430610>